File: match/practice/2014_11_2/a.py

n = int(raw_input())
k = int(raw_input())

M = []
for i in xrange(n+3):
  M.append([0]*(n+1)+[0,0])

for i in xrange(k):
  v, x, y = map(int, raw_input().split())
  M[x][y] = v

ans = 0
def move(d, v, xx, yy):
  global ans
  if d == 'L':
    for i in xrange(1, n+1):
      ind = 0
      p = 1
      flag = 1
      while p <= n:
        if flag:
          while p <= n and M[i][p] == 0:
            p += 1
          pre = M[i][p]
          p += 1
          flag = 0

        while p <= n and M[i][p] == 0:
          p += 1
          
        if pre == M[i][p] and pre:
          ind += 1
          M[i][ind] = pre<<1
          ans += pre<<1
          
          flag = 1
        elif pre:
          ind += 1
          M[i][ind] = pre
          pre = M[i][p]

        p += 1

      if not flag and pre:
        ind += 1
        M[i][ind] = pre

      for x in xrange(ind+1, n+1):
        M[i][x] = 0
  elif d == 'R':
    for i in xrange(1, n+1):
      ind = n+1
      p = n
      flag = 1
      while p:
        if flag:
          while p and M[i][p] == 0:
            p -= 1
          pre = M[i][p]
          if p:p -= 1
          flag = 0

        while p and M[i][p] == 0:
          p -= 1
        if pre == M[i][p] and pre:
          ind -= 1
          M[i][ind] = pre<<1
          ans += pre<<1
          flag = 1
        elif pre:
          ind -= 1
          M[i][ind] = pre
          pre = M[i][p]

        if p:p -= 1

      if not flag and pre:
        ind -= 1
        M[i][ind] = pre

      for x in xrange(1, ind):
        M[i][x] = 0
  elif d == 'U':
    for i in xrange(1, n+1):
      ind = 0
      p = 1
      flag = 1
      while p <= n:
        if flag:
          while p <= n and M[p][i] == 0:
            p += 1
          pre = M[p][i]
          p += 1
          flag = 0

        while p <= n and M[p][i] == 0:
          p += 1
        if pre == M[p][i] and pre:
          ind += 1
          M[ind][i] = pre<<1
          ans += pre<<1
          flag = 1
        elif pre:
          ind += 1
          M[ind][i] = pre
          pre = M[p][i]

        p += 1

      if not flag and pre:
        ind += 1
        M[ind][i] = pre

      for x in xrange(ind+1, n+1):
        M[x][i] = 0
  elif d == 'D':
    for i in xrange(1, n+1):
      ind = n+1
      p = n
      flag = 1
      while p:
        if flag:
          while p and M[p][i] == 0:
            p -= 1
          pre = M[p][i]
          if p:p -= 1
          flag = 0

        while p and M[p][i] == 0:
          p -= 1
        if pre == M[p][i] and pre:
          ind -= 1
          M[ind][i] = pre<<1
          ans += pre<<1
          flag = 1
        elif pre:
          ind -= 1
          M[ind][i] = pre
          pre = M[p][i]

        if p:p -= 1

      if not flag and pre:
        ind -= 1
        M[ind][i] = pre

      for x in xrange(1, ind):
        M[x][i] = 0
  M[xx][yy] = v


def out():
  print
  for i in xrange(1, n+1):
    for j in xrange(1, n+1):
      print M[i][j],
    print
  print

l = int(raw_input())
#out()
for i in xrange(l):
  d, v, x, y = raw_input().split()
  v, x , y = map(int, [v, x, y])
  move(d, v, x, y)
  #out()

print ans

File: match/practice/2014_11_2/a.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const int maxn = 110;
int M[maxn][maxn];
int n, k;
int v, x, y;


void init() {
  scanf("%d", &n);
  scanf("%d", &k);
  memset(M, 0, sizeof(M));
  for (int i  = 0;  i< k; ++ i) {
    scanf("%d %d %d", &v, &x, &y);
    M[x][y] = v;
  }
}

long long ans = 0;
void move(char d, int v, int xx, int yy) {
  if (d == 'L') {
    for (int i = 1; i <= n; ++ i) {
      int ind = 0;
      int p = 1;
      int flag = 1;
      int pre;
      while (p <= n) {
        if (flag) {
          while (p <= n && M[i][p] == 0)
            p += 1;
          pre = M[i][p];
          p += 1;
          flag = 0;
        }

        while (p <= n && M[i][p] == 0) 
          p += 1;
          
        if (pre == M[i][p] && pre) {
          ind += 1;
          M[i][ind] = pre<<1;
          ans += pre<<1;
          flag = 1;
        }
        else if( pre) {
          ind += 1;
          M[i][ind] = pre;
          pre = M[i][p];
        }

        p += 1;
      }

      if (!flag && pre) {
        ind += 1;
        M[i][ind] = pre;
      }
      for (int x = ind+1; x <= n; ++ x)
        M[i][x] = 0;
    }
  } else if (d == 'R') {
    for (int i = 1; i <= n; ++ i) {
      int ind = n+1;
      int p = n;
      int flag = 1;
      int pre;
      while (p) {
        if (flag) {
          while (p&& M[i][p] == 0)
            p -= 1;
          pre = M[i][p];
          if (p)p -= 1;
          flag = 0;
        }

        while (p&& M[i][p] == 0) 
          p -= 1;
          
        if (pre == M[i][p] && pre) {
          ind -= 1;
          M[i][ind] = pre<<1;
          ans += pre<<1;
          flag = 1;
        }
        else if( pre) {
          ind -= 1;
          M[i][ind] = pre;
          pre = M[i][p];
        }

        if (p) p -= 1;
      }

      if (!flag && pre) {
        ind -= 1;
        M[i][ind] = pre;
      }
      for (int x = 1; x < ind; ++ x)
        M[i][x] = 0;
    }
  } else  if (d == 'U') {
    for (int i = 1; i <= n; ++ i) {
      int ind = 0;
      int p = 1;
      int flag = 1;
      int pre;
      while (p <= n) {
        if (flag) {
          while (p <= n && M[p][i] == 0)
            p += 1;
          pre = M[p][i];
          p += 1;
          flag = 0;
        }

        while (p <= n && M[p][i] == 0) 
          p += 1;
          
        if (pre == M[p][i] && pre) {
          ind += 1;
          M[ind][i] = pre<<1;
          ans += pre<<1;
          flag = 1;
        }
        else if( pre) {
          ind += 1;
          M[ind][i] = pre;
          pre = M[p][i];
        }

        p += 1;
      }

      if (!flag && pre) {
        ind += 1;
        M[ind][i] = pre;
      }
      for (int x = ind+1; x <= n; ++ x)
        M[x][i] = 0;
    }
  } else if (d == 'D') {
    for (int i = 1; i <= n; ++ i) {
      int ind = n+1;
      int p = n;
      int flag = 1;
      int pre;
      while (p) {
        if (flag) {
          while (p&& M[p][i] == 0)
            p -= 1;
          pre = M[p][i];
          if (p)p -= 1;
          flag = 0;
        }

        while (p&& M[p][i] == 0) 
          p -= 1;
          
        if (pre == M[p][i] && pre) {
          ind -= 1;
          M[ind][i] = pre<<1;
          ans += pre<<1;
          flag = 1;
        }
        else if( pre) {
          ind -= 1;
          M[ind][i] = pre;
          pre = M[p][i];
        }

        if (p) p -= 1;
      }

      if (!flag && pre) {
        ind -= 1;
        M[ind][i] = pre;
      }
      for (int x = 1; x < ind; ++ x)
        M[x][i] = 0;
    }
  }
  M[xx][yy] = v;
}


//def out():
//  print
//  for i in xrange(1, n+1):
//    for j in xrange(1, n+1):
//      print M[i][j],
//    print
//  print

int main() {
  init();
  int l;
  scanf("%d", &l);
  char d[2];
  int v, x, y;
  for (int i = 0; i < l; ++ i) {
    scanf("%s %d %d %d", d, &v, &x, &y);
    move(d[0], v, x, y);
  }
  
  printf("%I64d\n", ans);
  return 0;
}
